File: ios/StockHODLShared/Live/PollPolicy.swift
import Foundation

/// The quote-poll gating policy, ported from `src/lib/holdings/poll-policy.ts`.
///
/// The distinction this carries — and that one derived boolean cannot — is
/// STRUCTURAL versus TRANSIENT emptiness:
///
///   - structural: the portfolio holds nothing the vendor can price. Polling
///     every ten seconds would fetch the same nulls forever;
///   - transient: the vendor hiccupped and one payload came back with every
///     price null. The cadence must keep running so the view self-heals.
///
/// Deciding "anything pollable?" from whether prices arrived conflates the two
/// and kills the cadence permanently on a single blip. The server already made
/// the structural call from the symbol set it actually sent to the vendor;
/// `hasPollableSymbols` on the payload IS that verdict, and this client
/// consumes it rather than second-guessing it from prices it can see.
///
/// The web equivalent of `hasPollableSymbols(outcomes)` has no counterpart here
/// on purpose: it runs over per-symbol vendor verdicts, which never leave the
/// server.
enum PollPolicy {
    /// Statuses during which the cadence runs — regular hours plus both
    /// extended sessions.
    static func isActiveStatus(_ status: MarketStatus) -> Bool {
        switch status {
        case .marketStatusOpen, .earlyTrading, .lateTrading:
            true
        case .closed, .unknown:
            false
        }
    }

    /// Poll while a session is running AND the portfolio structurally has
    /// something to poll for. Deliberately blind to whether the latest payload
    /// carried prices.
    static func shouldPollQuotes(status: MarketStatus, hasPollableSymbols: Bool) -> Bool {
        isActiveStatus(status) && hasPollableSymbols
    }

    /// Convenience over a whole payload, which is how every call site actually
    /// has the two facts to hand.
    static func shouldPoll(_ payload: LivePayload) -> Bool {
        shouldPollQuotes(
            status: payload.market.status,
            hasPollableSymbols: payload.hasPollableSymbols
        )
    }

    /// The cadence itself. Ten seconds matches the web client; the vendor feed
    /// is delayed anyway, so a tighter interval buys nothing but battery.
    static let interval: Duration = .seconds(10)
}

File: src/lib/holdings/poll-policy.ts
import type { Quote, QuoteOutcome } from '@/lib/market-data/provider';

/**
 * The quote-poll gating policy — pure and isomorphic (no `server-only`, no
 * fetch, no env) so the client component and the server builder share ONE
 * definition and the unit tests can import it directly.
 *
 * The point of this module is a distinction one derived boolean cannot carry:
 *
 * - STRUCTURAL: the portfolio holds nothing the vendor can price (e.g. only
 *   `CDR.WA`, which Massive answers with `not_found`). Polling every 10 s
 *   would fetch the same nulls forever — correctly skip the cadence.
 * - TRANSIENT: the vendor hiccupped (timeout / 500 / parse failure) and ONE
 *   payload came back with every price null. The cadence must keep running so
 *   the view self-heals on the next good poll.
 *
 * Deriving "anything pollable?" from whether prices came back conflates the
 * two and permanently kills the 10 s interval on a single blip — the client
 * gate therefore consumes only the server-computed structural verdict below,
 * which is derived from the symbol set actually SENT to the vendor and its
 * per-symbol verdicts, never from this round's prices.
 */

/** Statuses during which the quote cadence runs (regular + extended hours). */
const ACTIVE_STATUSES: ReadonlySet<Quote['marketStatus']> = new Set([
  'open',
  'early_trading',
  'late_trading',
]);

export function isActiveStatus(status: Quote['marketStatus']): boolean {
  return ACTIVE_STATUSES.has(status);
}

/**
 * The structural fact: does this portfolio hold anything the vendor might
 * price? `not_found` / `unsupported` are the vendor's *structural* verdicts
 * on a symbol; `ok` and `error` (timeout, HTTP failure, parse failure — all
 * transient by nature) both count as pollable. An empty outcome set (no
 * symbols at all) is structural absence.
 */
export function hasPollableSymbols(outcomes: Iterable<QuoteOutcome>): boolean {
  for (const outcome of outcomes) {
    if (outcome.ok) return true;
    if (outcome.reason !== 'not_found' && outcome.reason !== 'unsupported') return true;
  }
  return false;
}

/**
 * The client's interval gate: poll while a session is running AND the
 * portfolio structurally has something to poll for. Deliberately blind to
 * whether the latest payload carried prices — see the module doc.
 */
export function shouldPollQuotes(
  status: Quote['marketStatus'],
  hasPollable: boolean,
): boolean {
  return isActiveStatus(status) && hasPollable;
}
